src/matches.rs

```diff
@@ -0,0 +1,43 @@
+//! Pattern Matching macros.
+
+/// Returns true if an expression matches a pattern.
+///
+/// # Example
+///
+/// ```
+/// # #[macro_use] extern crate mac;
+///
+/// # fn main() {
+/// assert!(matches!(2, 1 | 2 | 3));
+/// assert!(matches!('x', 'a' ... 'z'));
+/// assert!(!matches!(Some(1), None));
+/// assert!(matches!(Some(42), Some(n) if n == 42));
+/// # }
+/// ```
+#[macro_export]
+macro_rules! matches {
+    ($expr:expr, $($pat:tt)+) => {
```

Comment:
It should be `$( $patter: pat )|+`, but that’s forbidden by https://github.com/rust-lang/rfcs/blob/master/text/0550-macro-future-proofing.md . This is why `_tt_as_expr_hack` exists.